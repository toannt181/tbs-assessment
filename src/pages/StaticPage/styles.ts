import styled from 'styled-components'

export const CatalogContainer = styled.div`
  padding-bottom: 32px;
`

export const Content = styled.div`
  max-width: 1000px;
  padding: 16px 32px;
  margin: auto;
`

export const SearchBox = styled.div`
  margin: 16px 0 32px;
`

export const Notice = styled.div`
  text-align: center;
  margin: 16px 0;
`
