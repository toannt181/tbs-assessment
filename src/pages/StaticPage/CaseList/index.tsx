import React, { memo, useRef, useEffect } from 'react'
import { CaseListContainer, Case, Label } from './styles'
import { ICase } from 'types/ICase'

interface ICaseListProps {
  cases: ICase[]
  onReachBottom: () => void
}

const CaseList = ({ cases, onReachBottom }: ICaseListProps) => {
  const containerRef = useRef<HTMLDivElement>(null)
  const loadingRef = useRef<HTMLDivElement>(null)

  useEffect(() => {
    if (!loadingRef.current || !containerRef.current) return
    const loadingDom = loadingRef.current as Element

    const callback = (entries: IntersectionObserverEntry[]) => {
      entries.forEach((entry) => {
        if (entry.isIntersecting) {
          onReachBottom()
        }
      })
    }
    const observer = new IntersectionObserver(callback)

    observer.observe(loadingDom)

    return () => observer.unobserve(loadingDom)
  }, [onReachBottom])

  return (
    <CaseListContainer ref={containerRef}>
      <ul data-testid="list">
        {cases.map((item) => (
          <Case key={item.id}>
            <div>
              <Label>Country</Label>
              {item.Country}
            </div>
            <div>
              <Label>CountryCode</Label>
              {item.CountryCode}
            </div>
            <div>
              <Label>Province</Label>
              {item.Province}
            </div>
            <div>
              <Label>Latitude</Label>
              {item.Lat}
            </div>
            <div>
              <Label>Longitude</Label>
              {item.Lon}
            </div>
            <div>
              <Label>Date</Label>
              {item.Date}
            </div>
          </Case>
        ))}
      </ul>
      <div ref={loadingRef} />
    </CaseListContainer>
  )
}

export default memo(CaseList)
