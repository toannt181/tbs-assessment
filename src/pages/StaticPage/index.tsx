import React, { useRef, useState, useEffect, useCallback, useMemo } from 'react'
import Header from 'components/Header'
import SearchableInput from 'components/SearchableInput'
import CaseList from './CaseList'
import { CatalogContainer, Content, SearchBox, Notice } from './styles'
import { ICase } from 'types/ICase'
import { IOption } from 'types/IOption'
import getCountries from 'api/getCountries'
import getCases from 'api/getCases'
import { DELAY_TIME, NUMBER_PER_PAGE } from 'constants/config'

const StaticPage = () => {
  const [keyword, setKeyword] = useState('')
  const [cases, setCases] = useState<ICase[]>([])
  const [options, setOptions] = useState<IOption[]>([])
  const [page, setPage] = useState(1)
  const timeoutNumber = useRef(0)

  useEffect(() => {
    const fetchCountries = async () => {
      const data: IOption[] = await getCountries()
      setOptions(data)
    }
    fetchCountries()
  }, [])

  useEffect(() => {
    if (timeoutNumber.current) {
      clearTimeout(timeoutNumber.current)
      timeoutNumber.current = 0
    }

    const fetchCases = async () => {
      const data: ICase[] = await getCases(keyword)
      setCases(data)
    }

    if (keyword) {
      timeoutNumber.current = setTimeout(fetchCases, DELAY_TIME)
    }

    return () => clearTimeout(timeoutNumber.current)
  }, [keyword])

  const limitedCases = useMemo(() => cases.slice(0, NUMBER_PER_PAGE * page), [page, cases])

  const onChangeKeyword = useCallback((value) => {
    setKeyword(value)
  }, [])

  const onReachBottom = useCallback(() => {
    if (page * NUMBER_PER_PAGE >= cases.length) return
    setPage((state) => state + 1)
  }, [page, cases])

  return (
    <CatalogContainer>
      <Header />
      <Content>
        <SearchBox>
          <SearchableInput
            value={keyword}
            options={options}
            keyword={keyword}
            onChangeKeyword={onChangeKeyword}
            placeholder="Search by country"
          />
        </SearchBox>
        {keyword && (
          <>
            <Notice>
              {cases.length === 0
                ? "Can't found Covid-19 Cases data"
                : `Found ${cases.length} cases, from 1 to ${Math.min(page * NUMBER_PER_PAGE, cases.length)}`}
            </Notice>
            {cases.length > 0 && <CaseList cases={limitedCases} onReachBottom={onReachBottom} />}
          </>
        )}
      </Content>
    </CatalogContainer>
  )
}

export default StaticPage
