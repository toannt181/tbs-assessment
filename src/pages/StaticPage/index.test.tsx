import StaticPage from '.'
import React from 'react'
import { render, fireEvent, screen } from '@testing-library/react'

describe('Static Page', () => {
  it('should render the static page', () => {
    const { container } = render(<StaticPage />)
    expect(container).toMatchSnapshot()
  })

  it('should render correct page', () => {
    render(<StaticPage />)
    expect(screen.getByText('Covid-19 Cases')).toBeTruthy()
  })

  it('should type value on input', () => {
    render(<StaticPage />)
    const input = screen.getByPlaceholderText('Search by country')
    fireEvent.change(input, { target: { value: 'example' } })
    expect(input.getAttribute('value')).toBe('example')
  })
})
