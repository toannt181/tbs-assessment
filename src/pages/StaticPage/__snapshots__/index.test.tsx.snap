// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Static Page should render the static page 1`] = `
<div>
  <div
    class="sc-pFZIQ riGkP"
  >
    <header
      class="sc-gsTCUz etYDuf"
    >
      <div
        class="sc-bdfBwQ gkIJkN"
      >
        Covid-19 Cases
      </div>
    </header>
    <div
      class="sc-jrAGrp cnVXXV"
    >
      <div
        class="sc-kEjbxe gLIwce"
      >
        <div
          class="sc-dlfnbm laaXGn"
        >
          <input
            class="sc-jSgupP gLwTsS"
            placeholder="Search by country"
            value=""
          />
        </div>
      </div>
      
    </div>
  </div>
</div>
`;
